File: frontend/src/types/report.ts
import type { PlanAnnotationType } from './plan'

export type Decision = 'A_BETTER' | 'B_BETTER' | 'BOTH_BAD' | 'BOTH_GOOD'
export type DistributionItem = {
  key: string
  count: number
}

export type AnnotationListParams = {
  operator_user_id?: number
  decision?: Decision
  result?: ManualCaseAnnotationResult
  start_date?: string
  end_date?: string
  page?: number
  page_size?: number
}

export type PlanStats = {
  plan_id: number
  annotation_type?: PlanAnnotationType
  total_cases: number
  annotated_cases: number
  pending_cases: number
  completion_rate: number
  decision_distribution?: Record<Decision, number>
  reason_distribution?: DistributionItem[]
  has_issues_cases?: number
  no_issue_cases?: number
}

export type AnnotationDetail = {
  id: number
  case_id: number
  hospitalization_no: string
  operator_user_id: number
  operator_username?: string | null
  decision: Decision
  reason_codes: string[]
  other_reason_text?: string | null
  notes?: string | null
  record_text: string
  agent_a_output: string
  agent_b_output: string
  display_a_source: string
  display_b_source: string
  created_at: string
}

export type AnnotationListResponse = {
  items: AnnotationDetail[]
  total: number
  page: number
  page_size: number
}

export type ManualCaseAnnotationResult = 'has_issues' | 'no_issue'

export type ManualCaseAnnotationSummary = {
  manual_annotation_id: number
  case_id: number
  hospitalization_no: string
  operator_user_id: number
  operator_username?: string | null
  result: ManualCaseAnnotationResult
  problem_count: number
  submitted_at: string
}

export type ManualAnnotationEntryDetail = {
  entry_id: number
  source_text: string
  start_offset: number
  end_offset: number
  quality_rule: {
    id: number
    category: string
    content: string
    score: string
  }
  suggestion: string
  notes?: string | null
  created_at: string
}

export type ManualAnnotationDetail = {
  manual_annotation_id: number
  case_id: number
  hospitalization_no: string
  operator: string
  result: ManualCaseAnnotationResult
  record_text: string
  submitted_at: string
  entries: ManualAnnotationEntryDetail[]
}

export type ManualAnnotationListResponse = {
  items: ManualCaseAnnotationSummary[]
  total: number
  page: number
  page_size: number
}

File: backend/app/schemas/report.py
from datetime import datetime

from pydantic import BaseModel

from app.models.entities import ManualAnnotationResult, PlanAnnotationType


class DistributionItem(BaseModel):
    key: str
    count: int


class PlanStatsResponse(BaseModel):
    plan_id: int
    annotation_type: PlanAnnotationType = PlanAnnotationType.COMPARISON
    total_cases: int
    annotated_cases: int
    pending_cases: int
    completion_rate: float
    decision_distribution: dict[str, int] | None = None
    reason_distribution: list[DistributionItem] | None = None
    has_issues_cases: int | None = None
    no_issue_cases: int | None = None


class AnnotationDetailItem(BaseModel):
    id: int
    case_id: int
    hospitalization_no: str
    operator_user_id: int
    operator_username: str | None = None
    decision: str
    reason_codes: list[str]
    other_reason_text: str | None = None
    notes: str | None = None
    record_text: str
    agent_a_output: str
    agent_b_output: str
    display_a_source: str
    display_b_source: str
    created_at: datetime


class AnnotationListResponse(BaseModel):
    items: list[AnnotationDetailItem]
    total: int
    page: int
    page_size: int


class ManualAnnotationSummaryItem(BaseModel):
    manual_annotation_id: int
    case_id: int
    hospitalization_no: str
    operator_user_id: int
    operator_username: str | None = None
    result: ManualAnnotationResult
    problem_count: int
    submitted_at: datetime


class ManualAnnotationListResponse(BaseModel):
    items: list[ManualAnnotationSummaryItem]
    total: int
    page: int
    page_size: int


class ManualAnnotationEntryDetail(BaseModel):
    entry_id: int
    source_text: str
    start_offset: int
    end_offset: int
    quality_rule: dict[str, int | str]
    suggestion: str
    notes: str | None = None
    created_at: datetime


class ManualAnnotationDetailResponse(BaseModel):
    manual_annotation_id: int
    case_id: int
    hospitalization_no: str
    operator: str
    result: ManualAnnotationResult
    record_text: str
    submitted_at: datetime
    entries: list[ManualAnnotationEntryDetail]
